x-pack/plugins/lens/public/editor_frame_plugin/editor_frame/editor_frame.tsx

```diff
@@ -49,27 +72,24 @@ export function EditorFrame(props: EditorFrameProps) {
     [state.datasource.activeId, state.datasource.isLoading]
   );
 
-  // create public datasource api for current state
-  // as soon as datasource is available and memoize it
-  const datasourcePublicAPI = useMemo(
-    () =>
-      state.datasource.activeId && !state.datasource.isLoading
-        ? props.datasourceMap[state.datasource.activeId].getPublicAPI(
-            state.datasource.state,
-            (newState: unknown) => {
-              dispatch({
-                type: 'UPDATE_DATASOURCE_STATE',
-                newState,
-              });
-            }
-          )
-        : undefined,
-    [
-      props.datasourceMap,
-      state.datasource.isLoading,
-      state.datasource.activeId,
-      state.datasource.state,
-    ]
+  // Initialize visualization as soon as datasource is ready
+  useEffect(
```

Comment:
This is fine, but I do find it hard to follow the chain of useEffects and how each affects the next. It might be nice to do all initialization in one place, and make it mostly linear: `initializeDatasource().then(ds => initializeVisualization(undefined, ds)).then(...)`

Obviously, not that ^^^ but something more direct and easier to reason about would be nice.

Reply:
I see what you mean. I did it like this to avoid creating the public datasource API in two separate places and flatten out the whole thing a bit.

We could initialize the datasource api separately and move everything into one effect like this:

```ts
  function getDatasourceAPI(datasourceState: unknown) {
    return props.datasourceMap[state.datasource.activeId!].getPublicAPI(
      datasourceState,
      (newState: unknown) => {
        dispatch({
          type: 'UPDATE_DATASOURCE_STATE',
          newState,
        });
      }
    );
  }

  // create public datasource api for current state
  // as soon as datasource is available and memoize it
  const datasourcePublicAPI = useMemo(
    () =>
      state.datasource.activeId && !state.datasource.isLoading
        ? getDatasourceAPI(state.datasource.state)
        : undefined,
    [
      props.datasourceMap,
      state.datasource.isLoading,
      state.datasource.activeId,
      state.datasource.state,
    ]
  );

  // Initialize current datasource
  useEffect(
    () => {
      let datasourceGotSwitched = false;
      if (state.datasource.isLoading && state.datasource.activeId) {
        props.datasourceMap[state.datasource.activeId].initialize().then(datasourceState => {
          if (!datasourceGotSwitched) {
            dispatch({
              type: 'UPDATE_DATASOURCE_STATE',
              newState: datasourceState,
            });

            if (state.visualization.state === null && state.visualization.activeId !== null) {
              const initialVisualizationState = props.visualizationMap[
                state.visualization.activeId
              ].initialize(getDatasourceAPI(datasourceState));
              dispatch({
                type: 'UPDATE_VISUALIZATION_STATE',
                newState: initialVisualizationState,
              });
            }
          }
        });

        return () => {
          datasourceGotSwitched = true;
        };
      }
    },
    [state.datasource.activeId, state.datasource.isLoading]
  );
```

But IMHO this doesn't make the flow much clearer. I'm going to merge it in the current state and we can discuss this today, but I suspect this init logic will change again when Wylie is going to look into the "switch visualization" flow.